apps/src/code-studio/components/header/HeaderFinish.jsx

```diff
@@ -0,0 +1,94 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import $ from 'jquery';
+
+const styles = {
+  headerContainer: {
+    position: 'relative',
+    overflow: 'hidden',
+    height: 40
+  },
+  headerInner: {
+    position: 'absolute'
+  },
+  scriptLinkWithUpdatedAt: {
+    display: 'block'
+  },
+  outerContainer: {
+    textAlign: 'right'
+  },
+  containerWithUpdatedAt: {
+    verticalAlign: 'bottom',
+    whiteSpace: 'nowrap',
+    overflow: 'hidden',
+    textOverflow: 'ellipsis',
+    display: 'inline-block'
+  },
+  headerVignetteRight: {
+    position: 'absolute',
+    width: '100%',
+    height: 18,
+    pointerEvents: 'none',
+    background:
+      'linear-gradient(to right, rgba(0, 173, 188, 0) calc(100% - 20px), rgba(0, 173, 188, 1) 100%)'
+  }
+};
+
+export default class HeaderFinish extends React.Component {
+  static propTypes = {
+    lessonData: PropTypes.object,
+    width: PropTypes.number,
+    setDesiredWidth: PropTypes.func
+  };
+
+  setDesiredWidth() {
+    // Report back to our parent how wide we would like to be.
+    const fullWidth = $('.header_finished').width();
+    if (this.props.setDesiredWidth) {
+      this.props.setDesiredWidth(fullWidth);
+    }
+  }
+
+  componentDidMount() {
+    this.setDesiredWidth();
+  }
+
+  componentDidUpdate() {
+    this.setDesiredWidth();
+  }
+
+  shouldComponentUpdate(nextProps, nextState) {
+    const lessonData = this.props.lessonData;
+
+    return (
+      this.props.width !== nextProps.width ||
+      !!lessonData !== !!nextProps.lessonData ||
+      lessonData.finishLink !== nextProps.lessonData.finishLink
+    );
+  }
+
+  render() {
+    const {lessonData} = this.props;
+
+    const fullWidth = $('.header_finished').width();
```

Comment:
Is it a bad idea to reach into the DOM on first render? This returns `null` in that case, but it seems like you've handled it fine.  Still... maybe surprising?